File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/core/MeteringPointFactoryImpl.kt
package dev.zeekr.camerax_android.core

import androidx.camera.core.MeteringPoint
import androidx.camera.core.MeteringPointFactory
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PigeonApiMeteringPointFactoryProxyApi

class MeteringPointFactoryImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiMeteringPointFactoryProxyApi(registrar) {
    override fun getDefaultPointSize(): Double {
        return MeteringPointFactory.getDefaultPointSize().toDouble()
    }

    override fun createPoint1(pigeon_instance: MeteringPointFactory, x: Double, y: Double): MeteringPoint {
        return pigeon_instance.createPoint(x.toFloat(), y.toFloat())
    }

    override fun createPoint2(
        pigeon_instance: MeteringPointFactory,
        x: Double,
        y: Double,
        size: Double
    ): MeteringPoint {
        return pigeon_instance.createPoint(x.toFloat(), y.toFloat(), size.toFloat())
    }
}

File: camerax_android/lib/src/core/metering_point_factory_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

abstract base class MeteringPointFactoryImpl implements MeteringPointFactory {
  MeteringPointFactoryProxyApi get api;

  @override
  Future<MeteringPoint> createPoint(double x, double y, {double? size}) =>
      size == null
      ? api.createPoint1(x, y).then((e) => e.impl)
      : api.createPoint2(x, y, size).then((e) => e.impl);
}

final class MeteringPointFactoryChannelImpl
    extends MeteringPointFactoryChannel {
  @override
  Future<double> getDefaultPointSize() =>
      MeteringPointFactoryProxyApi.getDefaultPointSize();
}
